diff --git a/llvm/lib/Support/BLAKE3/blake3_impl.h b/llvm/lib/Support/BLAKE3/blake3_impl.h
--- a/llvm/lib/Support/BLAKE3/blake3_impl.h
+++ b/llvm/lib/Support/BLAKE3/blake3_impl.h
@@ -32,16 +32,16 @@
 #define INLINE static inline __attribute__((always_inline))
 #endif
 
-#if defined(__x86_64__) || defined(_M_X64) 
+#if (defined(__x86_64__) || defined(_M_X64)) && !defined(__arm64ec__)
 #define IS_X86
 #define IS_X86_64
 #endif
 
 #if defined(__i386__) || defined(_M_IX86)
 #define IS_X86
 #define IS_X86_32
 #endif
 
-#if defined(__aarch64__) || defined(_M_ARM64)
+#if defined(__aarch64__) || defined(_M_ARM64) || defined(__arm64ec__)
 #define IS_AARCH64
 #endif
diff --git a/llvm/lib/Support/BLAKE3/CMakeLists.txt b/llvm/lib/Support/BLAKE3/CMakeLists.txt
--- a/llvm/lib/Support/BLAKE3/CMakeLists.txt
+++ b/llvm/lib/Support/BLAKE3/CMakeLists.txt
@@ -43,8 +43,9 @@
     endif()
   elseif(WIN32 OR CYGWIN)
     check_symbol_exists(__x86_64__ "" IS_X64)
-    if (IS_X64)
+    check_symbol_exists(__arm64ec__ "" IS_ARM64EC)
+    if (IS_X64 AND NOT IS_ARM64EC)
       list(APPEND LLVM_BLAKE3_FILES
         blake3_sse2_x86-64_windows_gnu.S
         blake3_sse41_x86-64_windows_gnu.S
         blake3_avx2_x86-64_windows_gnu.S
