diff --git a/llvm/lib/Support/Host.cpp b/llvm/lib/Support/Host.cpp
index 25ae1db..ccd30a5 100644
--- a/llvm/lib/Support/Host.cpp
+++ b/llvm/lib/Support/Host.cpp
@@ -380,5 +380,5 @@ StringRef sys::detail::getHostCPUNameForS390x(StringRef ProcCpuinfoContent) {

 StringRef sys::detail::getHostCPUNameForBPF() {
-#if !defined(__linux__) || !defined(__x86_64__)
+#if !defined(__linux__) || !(defined(__x86_64__) && !defined(__arm64ec__))
   return "generic";
 #else
@@ -445,6 +445,6 @@ StringRef sys::detail::getHostCPUNameForBPF() {
 }

-#if defined(__i386__) || defined(_M_IX86) || \
-    defined(__x86_64__) || defined(_M_X64)
+#if (defined(__i386__) || defined(_M_IX86) || \
+     defined(__x86_64__) || defined(_M_X64)) && !defined(__arm64ec__)

 // The check below for i386 was copied from clang's cpuid.h (__get_cpuid_max).
@@ -488,5 +488,5 @@ static bool getX86CpuIDAndInfo(unsigned value, unsigned *rEAX, unsigned *rEBX,
                                unsigned *rECX, unsigned *rEDX) {
 #if defined(__GNUC__) || defined(__clang__)
-#if defined(__x86_64__)
+#if (defined(__x86_64__) && !defined(__arm64ec__))
   // gcc doesn't know cpuid would clobber ebx/rbx. Preserve it manually.
   // FIXME: should we save this for Clang?
@@ -564,5 +564,5 @@ static bool getX86CpuIDAndInfoEx(unsigned value, unsigned subleaf,
                                  unsigned *rEDX) {
 #if defined(__GNUC__) || defined(__clang__)
-#if defined(__x86_64__)
+#if (defined(__x86_64__) && !defined(__arm64ec__))
   // gcc doesn't know cpuid would clobber ebx/rbx. Preserve it manually.
   // FIXME: should we save this for Clang?
@@ -1495,6 +1495,6 @@ int sys::getHostNumPhysicalCores() {
 }

-#if defined(__i386__) || defined(_M_IX86) || \
-    defined(__x86_64__) || defined(_M_X64)
+#if (defined(__i386__) || defined(_M_IX86) || \
+     defined(__x86_64__) || defined(_M_X64)) && !defined(__arm64ec__)
 bool sys::getHostCPUFeatures(StringMap<bool> &Features) {
   unsigned EAX = 0, EBX = 0, ECX = 0, EDX = 0;
