diff --git a/lldb/source/Host/common/File.cpp b/lldb/source/Host/common/File.cpp
index 15e7a21..bf24843 100644
--- a/lldb/source/Host/common/File.cpp
+++ b/lldb/source/Host/common/File.cpp
@@ -288,8 +288,9 @@ FILE *NativeFile::GetStream() {
           m_own_descriptor = true;
         }
 
-        m_stream = llvm::sys::RetryAfterSignal(nullptr, ::fdopen, m_descriptor,
-                                               mode.get());
+        m_stream = llvm::sys::RetryAfterSignal(nullptr, [&] {
+          return ::fdopen(m_descriptor, mode.get());
+        });
 
         // If we got a stream, then we own the stream and should no longer own
         // the descriptor because fclose() will close it for us
@@ -452,7 +453,7 @@ off_t NativeFile::SeekFromEnd(off_t offset, Status *error_ptr) {
 Status NativeFile::Flush() {
   Status error;
   if (StreamIsValid()) {
-    if (llvm::sys::RetryAfterSignal(EOF, ::fflush, m_stream) == EOF)
+    if (llvm::sys::RetryAfterSignal(EOF, [&] { return ::fflush(m_stream); }) == EOF)
       error.SetErrorToErrno();
   } else if (!DescriptorIsValid()) {
     error.SetErrorString("invalid file handle");
@@ -468,7 +469,7 @@ Status NativeFile::Sync() {
     if (err == 0)
       error.SetErrorToGenericError();
 #else
-    if (llvm::sys::RetryAfterSignal(-1, ::fsync, m_descriptor) == -1)
+    if (llvm::sys::RetryAfterSignal(-1, [&] { return ::fsync(m_descriptor); }) == -1)
       error.SetErrorToErrno();
 #endif
   } else {
@@ -518,7 +519,7 @@ Status NativeFile::Read(void *buf, size_t &num_bytes) {
 
   ssize_t bytes_read = -1;
   if (DescriptorIsValid()) {
-    bytes_read = llvm::sys::RetryAfterSignal(-1, ::read, m_descriptor, buf, num_bytes);
+    bytes_read = llvm::sys::RetryAfterSignal(-1, [&] { return ::read(m_descriptor, buf, num_bytes); });
     if (bytes_read == -1) {
       error.SetErrorToErrno();
       num_bytes = 0;
@@ -578,7 +579,7 @@ Status NativeFile::Write(const void *buf, size_t &num_bytes) {
   ssize_t bytes_written = -1;
   if (DescriptorIsValid()) {
     bytes_written =
-        llvm::sys::RetryAfterSignal(-1, ::write, m_descriptor, buf, num_bytes);
+        llvm::sys::RetryAfterSignal(-1, [&] { return ::write(m_descriptor, buf, num_bytes); });
     if (bytes_written == -1) {
       error.SetErrorToErrno();
       num_bytes = 0;
@@ -641,7 +642,7 @@ Status NativeFile::Read(void *buf, size_t &num_bytes, off_t &offset) {
   int fd = GetDescriptor();
   if (fd != kInvalidDescriptor) {
     ssize_t bytes_read =
-        llvm::sys::RetryAfterSignal(-1, ::pread, fd, buf, num_bytes, offset);
+        llvm::sys::RetryAfterSignal(-1, [&] { return ::pread(fd, buf, num_bytes, offset); });
     if (bytes_read < 0) {
       num_bytes = 0;
       error.SetErrorToErrno();
@@ -701,7 +702,7 @@ Status NativeFile::Write(const void *buf, size_t &num_bytes, off_t &offset) {
   if (fd != kInvalidDescriptor) {
 #ifndef _WIN32
     ssize_t bytes_written =
-        llvm::sys::RetryAfterSignal(-1, ::pwrite, m_descriptor, buf, num_bytes, offset);
+        llvm::sys::RetryAfterSignal(-1, [&] { return ::pwrite(m_descriptor, buf, num_bytes, offset); });
     if (bytes_written < 0) {
       num_bytes = 0;
       error.SetErrorToErrno();
diff --git a/lldb/source/Host/common/PseudoTerminal.cpp b/lldb/source/Host/common/PseudoTerminal.cpp
index be4c3c7..7a7e09d 100644
--- a/lldb/source/Host/common/PseudoTerminal.cpp
+++ b/lldb/source/Host/common/PseudoTerminal.cpp
@@ -95,7 +95,9 @@ llvm::Error PseudoTerminal::OpenSecondary(int oflag) {
   CloseSecondaryFileDescriptor();
 
   std::string name = GetSecondaryName();
-  m_secondary_fd = llvm::sys::RetryAfterSignal(-1, ::open, name.c_str(), oflag);
+  m_secondary_fd = llvm::sys::RetryAfterSignal(-1, [&] {
+    return ::open(name.c_str(), oflag);
+  });
   if (m_secondary_fd >= 0)
     return llvm::Error::success();
 
diff --git a/lldb/source/Host/common/Socket.cpp b/lldb/source/Host/common/Socket.cpp
index c2fb550..2d7ee38 100644
--- a/lldb/source/Host/common/Socket.cpp
+++ b/lldb/source/Host/common/Socket.cpp
@@ -365,10 +365,10 @@ NativeSocket Socket::AcceptSocket(NativeSocket sockfd, struct sockaddr *addr,
     flags |= SOCK_CLOEXEC;
   }
   NativeSocket fd = llvm::sys::RetryAfterSignal(
-      static_cast<NativeSocket>(-1), ::accept4, sockfd, addr, addrlen, flags);
+      static_cast<NativeSocket>(-1), [&] { return ::accept4(sockfd, addr, addrlen, flags); });
 #else
   NativeSocket fd = llvm::sys::RetryAfterSignal(
-      static_cast<NativeSocket>(-1), ::accept, sockfd, addr, addrlen);
+      static_cast<NativeSocket>(-1), [&] { return ::accept(sockfd, addr, addrlen); });
 #endif
   if (fd == kInvalidSocketValue)
     SetLastError(error);
diff --git a/lldb/source/Host/common/TCPSocket.cpp b/lldb/source/Host/common/TCPSocket.cpp
index 82b00ac..385b528 100644
--- a/lldb/source/Host/common/TCPSocket.cpp
+++ b/lldb/source/Host/common/TCPSocket.cpp
@@ -167,9 +167,10 @@ Status TCPSocket::Connect(llvm::StringRef name) {
 
     address.SetPort(host_port->port);
 
-    if (llvm::sys::RetryAfterSignal(-1, ::connect, GetNativeSocket(),
-                                    &address.sockaddr(),
-                                    address.GetLength()) == -1) {
+    if (llvm::sys::RetryAfterSignal(-1, [&] {
+          return ::connect(GetNativeSocket(), &address.sockaddr(),
+                           address.GetLength());
+        }) == -1) {
       Close();
       continue;
     }
diff --git a/lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp b/lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp
index dc5b249..e394f2c 100644
--- a/lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp
+++ b/lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp
@@ -510,7 +510,7 @@ ConnectionFileDescriptor::BytesAvailable(const Timeout<std::micro> &timeout,
           char c;
 
           ssize_t bytes_read =
-              llvm::sys::RetryAfterSignal(-1, ::read, pipe_fd, &c, 1);
+              llvm::sys::RetryAfterSignal(-1, [&] { return ::read(pipe_fd, &c, 1); });
           assert(bytes_read == 1);
           (void)bytes_read;
           switch (c) {
@@ -726,7 +726,7 @@ ConnectionStatus ConnectionFileDescriptor::ConnectFile(
 #if LLDB_ENABLE_POSIX
   std::string addr_str = s.str();
   // file:///PATH
-  int fd = llvm::sys::RetryAfterSignal(-1, ::open, addr_str.c_str(), O_RDWR);
+  int fd = llvm::sys::RetryAfterSignal(-1, [&] { return ::open(addr_str.c_str(), O_RDWR); });
   if (fd == -1) {
     if (error_ptr)
       error_ptr->SetErrorToErrno();
@@ -749,7 +749,7 @@ ConnectionStatus ConnectionFileDescriptor::ConnectFile(
     options.c_cc[VMIN] = 1;
     options.c_cc[VTIME] = 0;
 
-    llvm::sys::RetryAfterSignal(-1, ::tcsetattr, fd, TCSANOW, &options);
+    llvm::sys::RetryAfterSignal(-1, [&] { return ::tcsetattr(fd, TCSANOW, &options); });
   }
 
   m_io_sp = std::make_shared<NativeFile>(fd, File::eOpenOptionReadWrite, true);
@@ -776,7 +776,7 @@ ConnectionStatus ConnectionFileDescriptor::ConnectSerialPort(
     return eConnectionStatusError;
   }
 
-  int fd = llvm::sys::RetryAfterSignal(-1, ::open, path.str().c_str(), O_RDWR);
+  int fd = llvm::sys::RetryAfterSignal(-1, [&] { return ::open(path.str().c_str(), O_RDWR); });
   if (fd == -1) {
     if (error_ptr)
       error_ptr->SetErrorToErrno();
diff --git a/lldb/source/Host/posix/DomainSocket.cpp b/lldb/source/Host/posix/DomainSocket.cpp
index ddbd983..c17086f 100644
--- a/lldb/source/Host/posix/DomainSocket.cpp
+++ b/lldb/source/Host/posix/DomainSocket.cpp
@@ -79,8 +79,10 @@ Status DomainSocket::Connect(llvm::StringRef name) {
   m_socket = CreateSocket(kDomain, kType, 0, m_child_processes_inherit, error);
   if (error.Fail())
     return error;
-  if (llvm::sys::RetryAfterSignal(-1, ::connect, GetNativeSocket(),
-        (struct sockaddr *)&saddr_un, saddr_un_len) < 0)
+  if (llvm::sys::RetryAfterSignal(-1, [&] {
+        return ::connect(GetNativeSocket(), (struct sockaddr *)&saddr_un,
+                         saddr_un_len);
+      }) < 0)
     SetLastError(error);
 
   return error;
diff --git a/lldb/source/Host/posix/FileSystemPosix.cpp b/lldb/source/Host/posix/FileSystemPosix.cpp
index 26a266e..492cd19 100644
--- a/lldb/source/Host/posix/FileSystemPosix.cpp
+++ b/lldb/source/Host/posix/FileSystemPosix.cpp
@@ -73,9 +73,11 @@ Status FileSystem::ResolveSymbolicLink(const FileSpec &src, FileSpec &dst) {
 }
 
 FILE *FileSystem::Fopen(const char *path, const char *mode) {
-  return llvm::sys::RetryAfterSignal(nullptr, ::fopen, path, mode);
+  return llvm::sys::RetryAfterSignal(nullptr, [&] { return ::fopen(path, mode); });
 }
 
 int FileSystem::Open(const char *path, int flags, int mode) {
-  return llvm::sys::RetryAfterSignal(-1, ::open, path, flags, mode);
+  return llvm::sys::RetryAfterSignal(-1, [&] {
+    return ::open(path, flags, mode);
+  });
 }
diff --git a/lldb/source/Host/posix/LockFilePosix.cpp b/lldb/source/Host/posix/LockFilePosix.cpp
index cb9ca5c..c9a7d8e 100644
--- a/lldb/source/Host/posix/LockFilePosix.cpp
+++ b/lldb/source/Host/posix/LockFilePosix.cpp
@@ -27,7 +27,7 @@ static Status fileLock(int fd, int cmd, int lock_type, const uint64_t start,
   fl.l_pid = ::getpid();
 
   Status error;
-  if (llvm::sys::RetryAfterSignal(-1, ::fcntl, fd, cmd, &fl) == -1)
+  if (llvm::sys::RetryAfterSignal(-1, [&] { return ::fcntl(fd, cmd, &fl); }) == -1)
     error.SetErrorToErrno();
 
   return error;
diff --git a/lldb/source/Host/posix/MainLoopPosix.cpp b/lldb/source/Host/posix/MainLoopPosix.cpp
index b185c3d..308f944 100644
--- a/lldb/source/Host/posix/MainLoopPosix.cpp
+++ b/lldb/source/Host/posix/MainLoopPosix.cpp
@@ -229,7 +229,7 @@ MainLoopPosix::MainLoopPosix() : m_triggering(false) {
   m_read_fds.insert({trigger_pipe_fd, [trigger_pipe_fd](MainLoopBase &loop) {
                        char c;
                        ssize_t bytes_read = llvm::sys::RetryAfterSignal(
-                           -1, ::read, trigger_pipe_fd, &c, 1);
+                           -1, [&] { return ::read(trigger_pipe_fd, &c, 1); });
                        assert(bytes_read == 1);
                        UNUSED_IF_ASSERT_DISABLED(bytes_read);
                        // NB: This implicitly causes another loop iteration
diff --git a/lldb/source/Host/posix/PipePosix.cpp b/lldb/source/Host/posix/PipePosix.cpp
index bd311ad..a8d5f49 100644
--- a/lldb/source/Host/posix/PipePosix.cpp
+++ b/lldb/source/Host/posix/PipePosix.cpp
@@ -148,7 +148,7 @@ Status PipePosix::OpenAsReader(llvm::StringRef name,
     flags |= O_CLOEXEC;
 
   Status error;
-  int fd = llvm::sys::RetryAfterSignal(-1, ::open, name.str().c_str(), flags);
+  int fd = llvm::sys::RetryAfterSignal(-1, [&] { return ::open(name.str().c_str(), flags); });
   if (fd != -1)
     m_fds[READ] = fd;
   else
diff --git a/lldb/source/Host/posix/ProcessLauncherPosixFork.cpp b/lldb/source/Host/posix/ProcessLauncherPosixFork.cpp
index a148c11..16c0837 100644
--- a/lldb/source/Host/posix/ProcessLauncherPosixFork.cpp
+++ b/lldb/source/Host/posix/ProcessLauncherPosixFork.cpp
@@ -71,7 +71,7 @@ static void DisableASLR(int error_fd) {
 }
 
 static void DupDescriptor(int error_fd, const char *file, int fd, int flags) {
-  int target_fd = llvm::sys::RetryAfterSignal(-1, ::open, file, flags, 0666);
+  int target_fd = llvm::sys::RetryAfterSignal(-1, [&] { return ::open(file, flags, 0666); });
 
   if (target_fd == -1)
     ExitWithError(error_fd, "DupDescriptor-open");
@@ -288,8 +288,10 @@ ProcessLauncherPosixFork::LaunchProcess(const ProcessLaunchInfo &launch_info,
   do {
     pos += r;
     buf.resize_for_overwrite(pos + 100);
-    r = llvm::sys::RetryAfterSignal(-1, read, pipe.GetReadFileDescriptor(),
-                                    buf.begin() + pos, buf.size() - pos);
+    r = llvm::sys::RetryAfterSignal(-1, [&] {
+      return read(pipe.GetReadFileDescriptor(), buf.begin() + pos,
+                  buf.size() - pos);
+    });
   } while (r > 0);
   assert(r != -1);
 
@@ -299,7 +301,7 @@ ProcessLauncherPosixFork::LaunchProcess(const ProcessLaunchInfo &launch_info,
 
   error.SetErrorString(buf);
 
-  llvm::sys::RetryAfterSignal(-1, waitpid, pid, nullptr, 0);
+  llvm::sys::RetryAfterSignal(-1, [&] { return waitpid(pid, nullptr, 0); });
 
   return HostProcess();
 }
diff --git a/lldb/source/Plugins/Process/FreeBSD/NativeProcessFreeBSD.cpp b/lldb/source/Plugins/Process/FreeBSD/NativeProcessFreeBSD.cpp
index dbaa83f..4ba9bc4 100644
--- a/lldb/source/Plugins/Process/FreeBSD/NativeProcessFreeBSD.cpp
+++ b/lldb/source/Plugins/Process/FreeBSD/NativeProcessFreeBSD.cpp
@@ -68,7 +68,7 @@ NativeProcessFreeBSD::Factory::Launch(ProcessLaunchInfo &launch_info,
 
   // Wait for the child process to trap on its call to execve.
   int wstatus;
-  ::pid_t wpid = llvm::sys::RetryAfterSignal(-1, ::waitpid, pid, &wstatus, 0);
+  ::pid_t wpid = llvm::sys::RetryAfterSignal(-1, [&] { return ::waitpid(pid, &wstatus, 0); });
   assert(wpid == pid);
   (void)wpid;
   if (!WIFSTOPPED(wstatus)) {
@@ -748,7 +748,7 @@ void NativeProcessFreeBSD::SigchldHandler() {
   Log *log = GetLog(POSIXLog::Process);
   int status;
   ::pid_t wait_pid =
-      llvm::sys::RetryAfterSignal(-1, waitpid, GetID(), &status, WNOHANG);
+      llvm::sys::RetryAfterSignal(-1, [&] { return waitpid(GetID(), &status, WNOHANG); });
 
   if (wait_pid == 0)
     return;
@@ -834,8 +834,9 @@ Status NativeProcessFreeBSD::Attach() {
   int wstatus;
   // Need to use WALLSIG otherwise we receive an error with errno=ECHLD At this
   // point we should have a thread stopped if waitpid succeeds.
-  if ((wstatus = llvm::sys::RetryAfterSignal(-1, waitpid, m_pid, nullptr, 0)) <
-      0)
+  if ((wstatus = llvm::sys::RetryAfterSignal(-1, [&] {
+                    return waitpid(m_pid, nullptr, 0);
+                  })) < 0)
     return Status(errno, eErrorTypePOSIX);
 
   // Initialize threads and tracing status
@@ -971,7 +972,7 @@ void NativeProcessFreeBSD::MonitorClone(::pid_t child_pid, bool is_vfork,
 
   int status;
   ::pid_t wait_pid =
-      llvm::sys::RetryAfterSignal(-1, ::waitpid, child_pid, &status, 0);
+      llvm::sys::RetryAfterSignal(-1, [&] { return ::waitpid(child_pid, &status, 0); });
   if (wait_pid != child_pid) {
     LLDB_LOG(log,
              "waiting for pid {0} failed. Assuming the pid has "
diff --git a/lldb/source/Plugins/Process/Linux/NativeProcessLinux.cpp b/lldb/source/Plugins/Process/Linux/NativeProcessLinux.cpp
index a0e5d25..d2927cb 100644
--- a/lldb/source/Plugins/Process/Linux/NativeProcessLinux.cpp
+++ b/lldb/source/Plugins/Process/Linux/NativeProcessLinux.cpp
@@ -236,7 +236,7 @@ NativeProcessLinux::Factory::Launch(ProcessLaunchInfo &launch_info,
 
   // Wait for the child process to trap on its call to execve.
   int wstatus = 0;
-  ::pid_t wpid = llvm::sys::RetryAfterSignal(-1, ::waitpid, pid, &wstatus, 0);
+  ::pid_t wpid = llvm::sys::RetryAfterSignal(-1, [&] { return ::waitpid(pid, &wstatus, 0); });
   assert(wpid == pid);
   (void)wpid;
   if (!WIFSTOPPED(wstatus)) {
@@ -356,7 +356,7 @@ llvm::Expected<std::vector<::pid_t>> NativeProcessLinux::Attach(::pid_t pid) {
         }
 
         int wpid =
-            llvm::sys::RetryAfterSignal(-1, ::waitpid, tid, nullptr, __WALL);
+            llvm::sys::RetryAfterSignal(-1, [&] { return ::waitpid(tid, nullptr, __WALL); });
         // Need to use __WALL otherwise we receive an error with errno=ECHLD At
         // this point we should have a thread stopped if waitpid succeeds.
         if (wpid < 0) {
@@ -499,7 +499,7 @@ void NativeProcessLinux::WaitForCloneNotification(::pid_t pid) {
            "waiting for it to appear...",
            pid);
   ::pid_t wait_pid =
-      llvm::sys::RetryAfterSignal(-1, ::waitpid, pid, &status, __WALL);
+      llvm::sys::RetryAfterSignal(-1, [&] { return ::waitpid(pid, &status, __WALL); });
 
   // It's theoretically possible to get other events if the entire process was
   // SIGKILLed before we got a chance to check this. In that case, we'll just
@@ -1287,8 +1287,9 @@ NativeProcessLinux::Syscall(llvm::ArrayRef<uint64_t> args) {
     return std::move(Err);
 
   int status;
-  ::pid_t wait_pid = llvm::sys::RetryAfterSignal(-1, ::waitpid, thread.GetID(),
-                                                 &status, __WALL);
+  ::pid_t wait_pid = llvm::sys::RetryAfterSignal(-1, [&] {
+    return ::waitpid(thread.GetID(), &status, __WALL);
+  });
   if (wait_pid == -1) {
     return llvm::errorCodeToError(
         std::error_code(errno, std::generic_category()));
@@ -1875,7 +1876,7 @@ static std::optional<WaitStatus> HandlePid(::pid_t pid) {
 
   int status;
   ::pid_t wait_pid = llvm::sys::RetryAfterSignal(
-      -1, ::waitpid, pid, &status, __WALL | __WNOTHREAD | WNOHANG);
+      -1, [&] { return ::waitpid(pid, &status, __WALL | __WNOTHREAD | WNOHANG); });
 
   if (wait_pid == 0)
     return std::nullopt;
diff --git a/lldb/source/Plugins/Process/Linux/SingleStepCheck.cpp b/lldb/source/Plugins/Process/Linux/SingleStepCheck.cpp
index e7124ef..7338255 100644
--- a/lldb/source/Plugins/Process/Linux/SingleStepCheck.cpp
+++ b/lldb/source/Plugins/Process/Linux/SingleStepCheck.cpp
@@ -55,7 +55,7 @@ struct ChildDeleter {
     // Kill the child.
     kill(pid, SIGKILL);
     // Pick up the remains.
-    llvm::sys::RetryAfterSignal(-1, waitpid, pid, &status, __WALL);
+    llvm::sys::RetryAfterSignal(-1, [&] { return waitpid(pid, &status, __WALL); });
   }
 };
 
@@ -84,8 +84,9 @@ bool WorkaroundNeeded() {
   }
 
   int status;
-  ::pid_t wpid = llvm::sys::RetryAfterSignal(-1, waitpid,
-      child_pid, &status, __WALL);
+  ::pid_t wpid = llvm::sys::RetryAfterSignal(-1, [&] {
+    return waitpid(child_pid, &status, __WALL);
+  });
   if (wpid != child_pid || !WIFSTOPPED(status)) {
     LLDB_LOG(log, "waitpid() failed (status = {0:x}): {1}", status,
              Status(errno, eErrorTypePOSIX));
@@ -114,8 +115,9 @@ bool WorkaroundNeeded() {
       break;
     }
 
-    wpid = llvm::sys::RetryAfterSignal(-1, waitpid,
-        child_pid, &status, __WALL);
+    wpid = llvm::sys::RetryAfterSignal(-1, [&] {
+      return waitpid(child_pid, &status, __WALL);
+    });
     if (wpid != child_pid || !WIFSTOPPED(status)) {
       LLDB_LOG(log, "waitpid() failed (status = {0:x}): {1}", status,
                Status(errno, eErrorTypePOSIX));
diff --git a/lldb/source/Plugins/Process/NetBSD/NativeProcessNetBSD.cpp b/lldb/source/Plugins/Process/NetBSD/NativeProcessNetBSD.cpp
index 182eefb..1aee8e4 100644
--- a/lldb/source/Plugins/Process/NetBSD/NativeProcessNetBSD.cpp
+++ b/lldb/source/Plugins/Process/NetBSD/NativeProcessNetBSD.cpp
@@ -73,7 +73,7 @@ NativeProcessNetBSD::Factory::Launch(ProcessLaunchInfo &launch_info,
 
   // Wait for the child process to trap on its call to execve.
   int wstatus;
-  ::pid_t wpid = llvm::sys::RetryAfterSignal(-1, ::waitpid, pid, &wstatus, 0);
+  ::pid_t wpid = llvm::sys::RetryAfterSignal(-1, [&] { return ::waitpid(pid, &wstatus, 0); });
   assert(wpid == pid);
   (void)wpid;
   if (!WIFSTOPPED(wstatus)) {
@@ -774,8 +774,9 @@ Status NativeProcessNetBSD::GetFileLoadAddress(const llvm::StringRef &file_name,
 void NativeProcessNetBSD::SigchldHandler() {
   Log *log = GetLog(POSIXLog::Process);
   int status;
-  ::pid_t wait_pid = llvm::sys::RetryAfterSignal(-1, waitpid, GetID(), &status,
-                                                 WALLSIG | WNOHANG);
+  ::pid_t wait_pid = llvm::sys::RetryAfterSignal(-1, [&] {
+    return waitpid(GetID(), &status, WALLSIG | WNOHANG);
+  });
 
   if (wait_pid == 0)
     return;
@@ -858,8 +859,9 @@ Status NativeProcessNetBSD::Attach() {
   int wstatus;
   // Need to use WALLSIG otherwise we receive an error with errno=ECHLD At this
   // point we should have a thread stopped if waitpid succeeds.
-  if ((wstatus = llvm::sys::RetryAfterSignal(-1, waitpid, m_pid, nullptr,
-                                             WALLSIG)) < 0)
+  if ((wstatus = llvm::sys::RetryAfterSignal(-1, [&] {
+                    return waitpid(m_pid, nullptr, WALLSIG);
+                  })) < 0)
     return Status(errno, eErrorTypePOSIX);
 
   // Initialize threads and tracing status
@@ -1018,7 +1020,7 @@ void NativeProcessNetBSD::MonitorClone(::pid_t child_pid, bool is_vfork,
 
   int status;
   ::pid_t wait_pid =
-      llvm::sys::RetryAfterSignal(-1, ::waitpid, child_pid, &status, 0);
+      llvm::sys::RetryAfterSignal(-1, [&] { return ::waitpid(child_pid, &status, 0); });
   if (wait_pid != child_pid) {
     LLDB_LOG(log,
              "waiting for pid {0} failed. Assuming the pid has "
diff --git a/lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp b/lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp
index 2291856..b1e5ed0 100644
--- a/lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp
+++ b/lldb/source/Plugins/ScriptInterpreter/Python/PythonDataObjects.cpp
@@ -130,7 +130,7 @@ void StructuredPythonObject::Serialize(llvm::json::OStream &s) const {
 
 void PythonObject::Dump(Stream &strm) const {
   if (m_py_obj) {
-    FILE *file = llvm::sys::RetryAfterSignal(nullptr, ::tmpfile);
+    FILE *file = llvm::sys::RetryAfterSignal(nullptr, [&] { return ::tmpfile(); });
     if (file) {
       ::PyObject_Print(m_py_obj, file, 0);
       const long length = ftell(file);
diff --git a/lldb/tools/lldb-vscode/lldb-vscode.cpp b/lldb/tools/lldb-vscode/lldb-vscode.cpp
index b70f5b0..50f2b33 100644
--- a/lldb/tools/lldb-vscode/lldb-vscode.cpp
+++ b/lldb/tools/lldb-vscode/lldb-vscode.cpp
@@ -141,8 +141,9 @@ SOCKET AcceptConnection(int portno) {
       listen(sockfd, 5);
       socklen_t clilen = sizeof(cli_addr);
       newsockfd =
-          llvm::sys::RetryAfterSignal(static_cast<SOCKET>(-1), accept, sockfd,
-                                      (struct sockaddr *)&cli_addr, &clilen);
+          llvm::sys::RetryAfterSignal(static_cast<SOCKET>(-1), [&] {
+            return accept(sockfd, (struct sockaddr *)&cli_addr, &clilen);
+          });
       if (newsockfd < 0)
         if (g_vsc.log)
           *g_vsc.log << "error: accept (" << strerror(errno) << ")"
